We will be working on a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.
The data dictionary provided with data is as follows:

   - `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
   - `name` - Name of the car.
   - `seller` - Whether the seller is private or a dealer.
   - `offerTyp`e - The type of listing
   - `price` - The price on the ad to sell the car.
   - `abtest` - Whether the listing is included in an A/B test.
   - `vehicleType` - The vehicle Type.
   - `yearOfRegistration` - The year in which which year the car was first registered.
   - `gearbox` - The transmission type.
   - `powerPS` - The power of the car in PS.
   - `model` - The car model name.
   - `kilometer` - How many kilometers the car has driven.
   - `monthOfRegistration` - The month in which which year the car was first registered.
   - `fuelType` - What type of fuel the car uses.
   - `brand` - The brand of the car.
   - `notRepairedDamage` - If the car has a damage which is not yet repaired.
   - `dateCreated` - The date on which the eBay listing was created.
   - `nrOfPictures` - The number of pictures in the ad.
   - `postalCode`- The postal code for the location of the vehicle.
   - `lastSeenOnline` - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the included used car listings.

In [1]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

In [2]:
autos = pd.read_csv("D:/Developer/Data Science/Projects/CSV files/autos.csv", encoding = 'Latin-1')
autos.head(3)

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46

### Cleaning Column Names

In [3]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [4]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test', 'vehicle_type', 
                 'registration_year', 'gearbox', 'power', 'model', 'odometer', 'registration_month', 'fuel_type', 
                 'brand', 'unrepaired_damage', 'date_created', 'number_pictures', 'postal_code','last_seen']

In [5]:
len(autos) #371528 rows imported

371528

In [6]:
len(autos.columns) #20 columns imported

20

### Renaming column

In [7]:
autos.rename({'odometer':'odometer_km'}, axis = 1, inplace = True)

### Changing Type

In [8]:
for col in ['date_crawled', 'date_created', 'last_seen']:
    autos[col] = pd.to_datetime(autos[col])
for col in ['name', 'seller', 'offer_type', 'ab_test', 'vehicle_type', 'gearbox', 'model', 'fuel_type', 'power']:
    autos[col] = autos[col].astype('category')

zfill consider whole postal code equal to zero rather than considering it as a some number 

In [9]:
autos['postal_code'] = autos['postal_code'].astype('str').str.zfill(5)
autos['odometer_km'] = autos['odometer_km'].astype('float')
autos['price'] = autos['price'].astype('float')

### Initial Exploration and Cleaning

In [10]:
autos['unrepaired_damage'].unique()

array([nan, 'ja', 'nein'], dtype=object)

In [11]:
autos['number_pictures'].unique()

array([0], dtype=int64)

In [12]:
autos['number_pictures'].nunique()

1

In [13]:
autos.drop('number_pictures', axis = 1, inplace = True)

In [14]:
autos.drop('unrepaired_damage', axis = 1, inplace = True)

In [15]:
autos.drop('offer_type', axis = 1, inplace = True)

In [16]:
autos.drop('seller', axis = 1, inplace = True)

In [17]:
autos.head(3)

date_crawled                            name    price ab_test  \
0 2016-03-24 11:52:17                      Golf_3_1.6    480.0    test   
1 2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  18300.0    test   
2 2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"   9800.0    test   

  vehicle_type  registration_year    gearbox power  model  odometer_km  \
0          NaN               1993    manuell     0   golf     150000.0   
1        coupe               2011    manuell   190    NaN     125000.0   
2          suv               2004  automatik   163  grand     125000.0   

   registration_month fuel_type       brand date_created postal_code  \
0                   0    benzin  volkswagen   2016-03-24       70435   
1                   5    diesel        audi   2016-03-24       66954   
2                   8    diesel        jeep   2016-03-14       90480   

            last_seen  
0 2016-04-07 03:16:57  
1 2016-04-07 01:46:50  
2 2016-04-05 12:47:46

In [18]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 16 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   date_crawled        371528 non-null  datetime64[ns]
 1   name                371528 non-null  category      
 2   price               371528 non-null  float64       
 3   ab_test             371528 non-null  category      
 4   vehicle_type        333659 non-null  category      
 5   registration_year   371528 non-null  int64         
 6   gearbox             351319 non-null  category      
 7   power               371528 non-null  category      
 8   model               351044 non-null  category      
 9   odometer_km         371528 non-null  float64       
 10  registration_month  371528 non-null  int64         
 11  fuel_type           338142 non-null  category      
 12  brand               371528 non-null  object        
 13  date_created        371528 no

In [19]:
autos.describe(include='all')

date_crawled         name         price ab_test vehicle_type  \
count                371528       371528  3.715280e+05  371528       333659   
unique               280500       233531           NaN       2            8   
top     2016-03-24 14:49:47  Ford_Fiesta           NaN    test    limousine   
freq                      7          657           NaN  192585        95894   
first   2016-03-05 14:06:22          NaN           NaN     NaN          NaN   
last    2016-04-07 14:36:58          NaN           NaN     NaN          NaN   
mean                    NaN          NaN  1.729514e+04     NaN          NaN   
std                     NaN          NaN  3.587954e+06     NaN          NaN   
min                     NaN          NaN  0.000000e+00     NaN          NaN   
25%                     NaN          NaN  1.150000e+03     NaN          NaN   
50%                     NaN          NaN  2.950000e+03     NaN          NaN   
75%                     NaN          NaN  7.200000e+03     NaN          NaN   
max                     NaN          NaN  2.147484e+09     NaN          NaN   

        registration_year  gearbox     power   model    odometer_km  \
count       371528.000000   351319  371528.0  351044  371528.000000   
unique                NaN        2     794.0     251            NaN   
top                   NaN  manuell       0.0    golf            NaN   
freq                  NaN   274214   40820.0   30070            NaN   
first                 NaN      NaN       NaN     NaN            NaN   
last                  NaN      NaN       NaN     NaN            NaN   
mean          2004.577997      NaN       NaN     NaN  125618.688228   
std             92.866598      NaN       NaN     NaN   40112.337051   
min           1000.000000      NaN       NaN     NaN    5000.000000   
25%           1999.000000      NaN       NaN     NaN  125000.000000   
50%           2003.000000      NaN       NaN     NaN  150000.000000   
75%           2008.000000      NaN       NaN     NaN  150000.000000   
max           9999.000000      NaN       NaN     NaN  150000.000000   

        registration_month fuel_type       brand         date_created  \
count        371528.000000    338142      371528               371528   
unique                 NaN         7          40                  114   
top                    NaN    benzin  volkswagen  2016-04-03 00:00:00   
freq                   NaN    223857       79640                14450   
first                  NaN       NaN         NaN  2014-03-10 00:00:00   
last                   NaN       NaN         NaN  2016-04-07 00:00:00   
mean              5.734445       NaN         NaN                  NaN   
std               3.712412       NaN         NaN                  NaN   
min               0.000000       NaN         NaN                  NaN   
25%               3.000000       NaN         NaN                  NaN   
50%               6.000000       NaN         NaN                  NaN   
75%               9.000000       NaN         NaN                  NaN   
max              12.000000       NaN         NaN                  NaN   

       postal_code            last_seen  
count       371528               371528  
unique        8150               182806  
top          10115  2016-04-06 13:45:54  
freq           828                   17  
first          NaN  2016-03-05 14:15:08  
last           NaN  2016-04-07 14:58:51  
mean           NaN                  NaN  
std            NaN                  NaN  
min            NaN                  NaN  
25%            NaN                  NaN  
50%            NaN                  NaN  
75%            NaN                  NaN  
max            NaN                  NaN

### Finding Outliers

***Find outliers with the Interquartile Range where q1 is 25 percentile and q3 75 percentile value of .describe()*** 

In [20]:
autos['price'].describe().apply(lambda x: format(x, 'f')) #scientific notation suppressed.

count        371528.000000
mean          17295.141865
std         3587953.744410
min               0.000000
25%            1150.000000
50%            2950.000000
75%            7200.000000
max      2147483647.000000
Name: price, dtype: object

In [21]:
autos['odometer_km'].describe().apply(lambda x: format(x, 'f'))

count    371528.000000
mean     125618.688228
std       40112.337051
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: object

In [22]:
cols = ['price', 'odometer_km']
q1 = autos[cols].quantile(0.25)
q3 = autos[cols].quantile(0.75)
iqr = q3 - q1

# Return a boolean array of the rows with (any) non-outlier column values
condition = ~((autos[cols] < (q1 - 1.5 * iqr)) | (autos[cols] > (q3 +1.5 * iqr))).any(axis = 1)

# Filter our dataframe based on condition
filtered_autos = autos[condition]

In [23]:
len(filtered_autos)

295620

**Original Rows: 371528**\
**After outliers**\
**Adjusted Rows: 295620**

### Exploring the Odometer and Price Columns

In [24]:
filtered_autos['odometer_km'].value_counts()

150000.0    235295
125000.0     35256
100000.0     14196
90000.0      10873
Name: odometer_km, dtype: int64

It can be seen that the values in this field are rounded, which might indicate that sellers had to choose from pre-set options for this field. Additionally, there are more high mileage than low mileage vehicles.

In [25]:
filtered_autos['price'].describe()

count    295620.000000
mean       3563.428415
std        3532.760660
min           0.000000
25%         999.000000
50%        2300.000000
75%        4999.000000
max       16270.000000
Name: price, dtype: float64

In [26]:
print('Total Number of Unique Values:',filtered_autos['price'].nunique())
print('Lowest Price:',filtered_autos['price'].min())
print('Highest Price:',filtered_autos['price'].max())

Total Number of Unique Values: 2990
Lowest Price: 0.0
Highest Price: 16270.0


Removing values which display no price.

In [27]:
filtered_price = filtered_autos[filtered_autos['price'].between(1, 16270)]

### Exploring the Date Columns

In [28]:
#showing distribution of values for dates
#to include missing values in the distribution and to use percentile instead of counts(dropna).
filtered_autos[['date_crawled','date_created','last_seen',]].value_counts(normalize = True).sort_index().head(10)

date_crawled         date_created  last_seen          
2016-03-05 14:06:22  2016-03-05    2016-03-27 17:16:15    0.000003
2016-03-05 14:06:23  2016-03-05    2016-03-06 18:29:35    0.000003
2016-03-05 14:06:24  2016-03-05    2016-03-09 15:19:01    0.000003
                                   2016-03-24 09:16:44    0.000003
                                   2016-04-06 22:15:33    0.000003
2016-03-05 14:06:25  2016-03-05    2016-03-09 15:19:01    0.000003
2016-03-05 14:06:26  2016-03-05    2016-03-05 17:47:48    0.000003
2016-03-05 14:06:27  2016-03-05    2016-03-21 02:18:33    0.000003
                                   2016-04-06 22:15:32    0.000003
2016-03-05 14:06:28  2016-03-05    2016-03-07 10:47:00    0.000003
dtype: float64

In [29]:
autos['registration_year'].describe()

count    371528.000000
mean       2004.577997
std          92.866598
min        1000.000000
25%        1999.000000
50%        2003.000000
75%        2008.000000
max        9999.000000
Name: registration_year, dtype: float64

In [30]:
filtered_autos['registration_year'].describe()

count    295620.000000
mean       2002.499144
std          14.918346
min        1000.000000
25%        1999.000000
50%        2002.000000
75%        2006.000000
max        9000.000000
Name: registration_year, dtype: float64

### Dealing with Incorrect Registration Year Data
A car can't be first registered before the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

The above mentioned problem can be solved by remove the listings with these values. Let's determine what percentage of our data has invalid values in this column:

In [31]:
filtered_autos_1900_2016 = filtered_autos[filtered_autos['registration_year'].between(1900, 2016)]

In [32]:
filtered_autos_1900_2016['registration_year'].value_counts(normalize = True)

2000    0.080238
1999    0.076699
2005    0.070608
2001    0.067229
2003    0.065698
          ...   
1930    0.000004
1928    0.000004
1919    0.000004
1946    0.000004
1915    0.000004
Name: registration_year, Length: 84, dtype: float64

It appears that most of the vehicles were first registered in the past 20 years.
### Exploring Price & Mileage by Brand

In [33]:
filtered_autos['brand'].value_counts()

volkswagen        66008
opel              34504
bmw               32704
mercedes_benz     27821
audi              25586
ford              20725
renault           15392
peugeot            9112
fiat               7373
seat               5517
mazda              4757
citroen            4075
skoda              4073
nissan             3924
toyota             3515
smart              3284
volvo              2950
mitsubishi         2561
hyundai            2352
honda              2338
alfa_romeo         2004
mini               1912
sonstige_autos     1822
kia                1742
suzuki             1584
chrysler           1287
chevrolet          1064
daihatsu            645
subaru              643
jeep                573
porsche             496
saab                492
daewoo              460
rover               440
land_rover          439
dacia               424
jaguar              421
lancia              403
trabant             120
lada                 78
Name: brand, dtype: int64

Selecting top 20 brands.

In [34]:
first20_brands = filtered_autos['brand'].value_counts().index[0:20]
first20_brands

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'mazda', 'citroen', 'skoda', 'nissan',
       'toyota', 'smart', 'volvo', 'mitsubishi', 'hyundai', 'honda'],
      dtype='object')

Below we will construct a dictionary with unique brand name as key and mean value of price and mileage as value. 

In [49]:
mean_price = {}
mean_mileage = {}
for key in first20_brands:
    selected_top20_brands = filtered_price[filtered_price['brand'] == key]
    mean_price[key] = round(selected_top20_brands['price'].agg(np.mean), 4)
    mean_mileage[key] = round(selected_top20_brands['odometer_km'].mean(), 4)
    
mean_price

{'volkswagen': 3634.3789,
 'opel': 2226.4184,
 'bmw': 5485.5927,
 'mercedes_benz': 5036.702,
 'audi': 5326.541,
 'ford': 2496.8949,
 'renault': 1803.18,
 'peugeot': 2459.514,
 'fiat': 1947.2039,
 'seat': 2911.239,
 'mazda': 2887.0002,
 'citroen': 2791.0335,
 'skoda': 4715.1214,
 'nissan': 3030.5123,
 'toyota': 3931.7494,
 'smart': 2414.3242,
 'volvo': 3721.7877,
 'mitsubishi': 2414.4399,
 'hyundai': 3504.8822,
 'honda': 2969.0408}

In [50]:
mean_mileage

{'volkswagen': 143724.7233,
 'opel': 142004.7475,
 'bmw': 145293.2935,
 'mercedes_benz': 144821.2711,
 'audi': 145807.7776,
 'ford': 140371.8516,
 'renault': 141260.3015,
 'peugeot': 139803.8337,
 'fiat': 137934.6607,
 'seat': 140673.1488,
 'mazda': 140903.4841,
 'citroen': 138841.8933,
 'skoda': 137866.2024,
 'nissan': 138844.5378,
 'toyota': 138216.763,
 'smart': 128980.7633,
 'volvo': 146385.9833,
 'mitsubishi': 141685.4839,
 'hyundai': 134242.0961,
 'honda': 141263.8581}

Converting from dictionary to dataframe.

In [77]:
def top20_brands(x, y):
    
    mean_price_df = pd.DataFrame.from_dict(x, orient = 'index', columns = ['mean_price'])
    mean_mileage_df = pd.DataFrame.from_dict(y, orient = 'index', columns = ['mean_mileage'])
    
    mean_price_mileage = pd.concat([mean_price_df, mean_mileage_df], axis = 1)
    
    return mean_price_mileage
    
top20_brands(mean_price, mean_mileage)

mean_price  mean_mileage
volkswagen      3634.3789   143724.7233
opel            2226.4184   142004.7475
bmw             5485.5927   145293.2935
mercedes_benz   5036.7020   144821.2711
audi            5326.5410   145807.7776
ford            2496.8949   140371.8516
renault         1803.1800   141260.3015
peugeot         2459.5140   139803.8337
fiat            1947.2039   137934.6607
seat            2911.2390   140673.1488
mazda           2887.0002   140903.4841
citroen         2791.0335   138841.8933
skoda           4715.1214   137866.2024
nissan          3030.5123   138844.5378
toyota          3931.7494   138216.7630
smart           2414.3242   128980.7633
volvo           3721.7877   146385.9833
mitsubishi      2414.4399   141685.4839
hyundai         3504.8822   134242.0961
honda           2969.0408   141263.8581

## Conclusion

We have utilized pandas and NumPy to quickly and easily clean, sort, and manipulate a large set of data. There were many things that required cleaning, as well as things that were technically fine, but needed to be changed for improved readability. We made heavy use of DataFrames in this project, as well as boolean index masking and aggregation.